cmd/snap-bootstrap/partition/disks.go

```diff
@@ -0,0 +1,499 @@
+			// bad udev output?
+			return nil, fmt.Errorf("cannot find disk for partition %s, bad udev output: %v", mountpointPart.path, err)
+		}
+		d.major = maj
+		d.minor = min
+	} else {
+		// didn't find the property we need
+		return nil, fmt.Errorf("cannot find disk for partition %s, incomplete udev output", mountpointPart.path)
+	}
+
+	// now we have the major and minor of the disk, and thus now have the
+	// arduous task to identify all partitions that come from this disk using
+	// sysfs
+	// step 1. find all devices with a matching major number
+	// step 2. start at the major + minor device for the disk, and iterate over
+	//         all devices that have a partition attribute, starting with the
+	//         device with major same as disk and minor equal to disk minor + 1
+	// step 3. if we hit a device that does not have a partition attribute, then
+	//         we hit another disk, and shall stop searching
+
+	// TODO: are there devices that have structures on them that show up as
+	//       contiguous devices but are _not_ partitions, i.e. some littlekernel
+	//       devices?
+
+	pattern := fmt.Sprintf("%s/dev/block/%d:*", dirs.SysfsDir, d.major)
+	allDevices, err := filepath.Glob(pattern)
+	if err != nil {
+		return nil, fmt.Errorf("internal error: %v", err)
+	}
+
+	// glob does not guarantee a sort, but we need our list of devices sorted
+	sort.Strings(allDevices)
+
+	// populate all the partitions from our candidate devices
+	for _, dev := range allDevices {
+		base := filepath.Base(dev)
+		maj, min, err := parseDeviceMajorMinor(base)
+		if err != nil {
+			continue
+		}
+
+		// ignore any devices that have minor numbers less than the disk itself
+		// the disk will have the same minor so ignore that too
+		if min <= d.minor {
+			continue
+		}
+
+		// now if there is a partition file, this is a partition for our disk
+		if osutil.FileExists(filepath.Join(dev, "partition")) {
+			// read the uevent file for this partition to get the devname
+			p := &partition{
+				major: maj,
+				minor: min,
+			}
+			f, err := os.Open(filepath.Join(dev, "uevent"))
+			if err != nil {
+				continue
+			}
+
+			// now get the full set of udev properties for this partition
+			eventProps, err := parseUdevProperties(f)
+			if err != nil {
+				continue
+			}
+
+			// get the name of the device path to call udevadm
+			// note that here DEVNAME is coming from sysfs's uevent, and for
+			// whatever reason it is just the basename a la vda3, but DEVNAME
+			// when  requested from udevadm proper is the full path a la
+			// /dev/vda3
+			if name, ok := eventProps["DEVNAME"]; ok {
+				p.path = filepath.Join("/dev", name)
+			} else {
+				continue
+			}
+
+			props, err := udevProperties(p.path)
+			if err != nil && props == nil {
+				// only error here if we didn't get a map, we validate the map
+				// in the next steps
+				continue
+			}
+
+			// get the label
+			if labelEncoded, ok := props["ID_FS_LABEL_ENC"]; ok {
+				p.label = labelEncoded
```

Comment:
so this is where the encoded thing come from that confused me? We probably need to either make it transparent as I said, or clarify and sprinkle "encoded" suffixed a bit more